File: dawn_go/surf.go
package surfbase

type node map[rune]node

func Build(keys []string) node {
    root := make(node)
    for _, key := range keys {
        curr := root
        for _, char := range key {
            next, ok := curr[char]
            if !ok {
                next = make(node)
                curr[char] = next
            }
            curr = next
        }
    }
    return root
}

func Lookup(trie node, key string) bool {
    curr := trie
    for _, char := range key {
        next, ok := curr[char]
        if !ok {
            return false
        }
        curr = next
    }
    return true
}

File: dawn_cpp/surf.cpp
#include <unordered_map>
#include <string>
#include <vector>

using Node = std::unordered_map<char, Node>;

Node build(const std::vector<std::string>& keys) {
    Node root;
    for (const auto& key : keys) {
        auto curr = &root;
        for (char c : key) {
            auto it = curr->find(c);
            if (it == curr->end()) {
                curr->emplace(c, Node());
                curr = &curr->at(c);
            } else {
                curr = &it->second;
            }
        }
    }
    return root;
}

bool lookup(const Node& trie, const std::string& key) {
    const Node* curr = &trie;
    for (char c : key) {
        auto it = curr->find(c);
        if (it == curr->end()) {
            return false;
        }
        curr = &it->second;
    }
    return true;
}